libdill.h

```diff
@@ -967,6 +967,29 @@ DILL_EXPORT int dill_ws_response_key(
 #define ws_response_key dill_ws_response_key
 #endif
 
+/******************************************************************************/
+/*  SOCKS5                                                                    */
+/******************************************************************************/
+
+struct dill_socks5_storage {char _[1200];};
+
+DILL_EXPORT int dill_socks5_attach_client(
+    int s, const char *username, const char *password, int64_t deadline);
+DILL_EXPORT int dill_socks5_attach_client_mem(
+    int s, const char *username, const char *password,
+    struct dill_socks5_storage *mem, int64_t deadline);
+DILL_EXPORT int dill_socks5_connect(
```

Comment:
Merge attach & connect functions. The design of libdill socks is done in such a way as to eliminate different intermediate states (e.g. socket is attached but not connected). The attach function should either fail or return a fully connected socket ready to be used.

Reply:
There are two steps to the protocol here:

1. Establish a connection (and authenticate if needed) to the proxy
2. One of the three possible commands to the proxy:
    a) CONNECT, which establishes an outbound TCP connection
    b) BIND, which causes the proxy to listen for incoming remote connections
    c) UDP ASSOCIATE, which creates a bidirectional UDP tunnel through the proxy. 

So far I've only implemented CONNECT. I modeled the API on the protocol here but could certainly combine the calls for at least the first two. 

UDP ASSOCIATE is a bit strange in that regard as the client first opens a TCP connection and sends a UDP ASSOCIATE command. That TCP connection is kept open (essentially a simple signalling channel). The client may then send UDP traffic through the proxy or listen for UDP messages which are routed through the proxy. When the TCP connection is closed the proxy removes the forwarding rules for that UDP tunnel. So I suppose the TCP connection would be a "fully connected socket ready to be used" so long as used equals "closed when done". Is that keeping with the design principle?

Reply:
Cases a) and b) could be solved the same way as, say, current implementation of TLS is in libdill:

```
sock5_attach_client()
sock5_attach_server()
```

c) is a weird case though. Sound almost as a different protocol (sock5_udp_...?) Anyway, what about solving a) and b) first, then try to figure out c) ?